rclcpp/src/rclcpp/node_interfaces/node_graph.cpp

```diff
@@ -369,3 +369,87 @@ NodeGraph::count_graph_users()
 {
   return graph_users_count_.load();
 }
+
+static
+std::vector<rclcpp::TopicEndpointInfo>
+convert_to_topic_info_list(const rmw_topic_endpoint_info_array_t & info_array)
+{
+  std::vector<rclcpp::TopicEndpointInfo> topic_info_list;
+  for (size_t i = 0; i < info_array.count; ++i) {
+    topic_info_list.push_back(rclcpp::TopicEndpointInfo(info_array.info_array[i]));
+  }
+  return topic_info_list;
+}
+
+template <const char * EndpointType, typename FunctionT>
+static std::vector<rclcpp::TopicEndpointInfo>
+get_info_by_topic(
+  rclcpp::node_interfaces::NodeBaseInterface * node_base,
+  const std::string & topic_name,
+  bool no_mangle,
+  FunctionT rcl_get_info_by_topic)
+{
+  std::vector<rclcpp::TopicEndpointInfo> topic_info_list;
+
+  auto rcl_node_handle = node_base->get_rcl_node_handle();
+  auto fqdn = rclcpp::expand_topic_or_service_name(
+    topic_name,
+    rcl_node_get_name(rcl_node_handle),
+    rcl_node_get_namespace(rcl_node_handle),
+    false);    // false = not a service
+
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  rmw_topic_endpoint_info_array_t info_array = rmw_get_zero_initialized_topic_endpoint_info_array();
+  auto ret =
+    rcl_get_info_by_topic(rcl_node_handle, &allocator, fqdn.c_str(), no_mangle, &info_array);
+  if (ret != RCL_RET_OK) {
+    auto error_msg =
+      std::string("Failed to get information by topic for ") + EndpointType + std::string(":");
+    if (ret == RCL_RET_UNSUPPORTED) {
+      error_msg += std::string("function not supported by RMW_IMPLEMENTATION");
+    } else {
+      error_msg += rcl_get_error_string().str;
+    }
+    rcl_reset_error();
+    if (rmw_topic_endpoint_info_array_fini(&info_array, &allocator) != RMW_RET_OK) {
+      error_msg += std::string(", failed also to cleanup topic info array, leaking memory: ")
+        + rcl_get_error_string().str;
+      rcl_reset_error();
+    }
+    throw_from_rcl_error(ret, error_msg);
+  }
+
+  topic_info_list = convert_to_topic_info_list(info_array);
+  ret = rmw_topic_endpoint_info_array_fini(&info_array, &allocator);
+  if (ret != RMW_RET_OK) {
+    throw_from_rcl_error(ret, "rmw_topic_info_array_fini failed.");
+  }
+
+  return topic_info_list;
+}
+
+static const char kPublisherEndpointTypeName[] = "publishers";
+std::vector<rclcpp::TopicEndpointInfo>
+NodeGraph::get_publishers_info_by_topic(
+  const std::string & topic_name,
+  bool no_mangle) const
+{
+  return get_info_by_topic<kPublisherEndpointTypeName>(
```

Comment:
why not just `get_info_by_topic<"publishers">`.

Reply:
The compiler will complain about lack of external linkage.

Reply:
Didn't know that that wasn't supported, thanks for clarifying!